src/main/java/ca/uhn/fhir/jpa/starter/cohort/service/r5/CohorteEvaluation.java

```diff
@@ -0,0 +1,69 @@
+package ca.uhn.fhir.jpa.starter.cohort.service.r5;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.hl7.elm.r1.ExpressionDef;
+import org.hl7.fhir.r5.model.EvidenceVariable;
+import org.hl7.fhir.r5.model.Group;
+import org.hl7.fhir.r5.model.Reference;
+import org.hl7.fhir.r5.model.ResearchStudy;
+import org.opencds.cqf.cql.engine.execution.CqlEngine;
+import org.opencds.cqf.cql.engine.execution.Libraries;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public class CohorteEvaluation {
+	private static final Logger logger = LoggerFactory.getLogger(CohorteEvaluation.class);
+	private final CqlEngine context;
+
+	public CohorteEvaluation(CqlEngine context) {
+		this.context = context;
+	}
+
+	protected Group evaluate(ResearchStudy researchStudy, EvidenceVariable evidenceVariable, List<String> subjectIds, Group group) {
+		logger.info("Evaluating Cohort {} with {} subject(s)", researchStudy.getUrl(), subjectIds.size());
+
+		for (String subjectId : subjectIds) {
+			if (subjectId == null) {
+				throw new RuntimeException("SubjectId is required in order to calculate.");
+			}
+
+			Pair<String, String> subjectInfo = this.getSubjectTypeAndId(subjectId);
+			String subjectTypePart = subjectInfo.getLeft();
+			String subjectIdPart = subjectInfo.getRight();
+			this.context.getState().setContextValue(subjectTypePart, subjectIdPart);
+			this.evaluateVariableDefinition(group, evidenceVariable, subjectTypePart, subjectIdPart);
+		}
+
+		return group;
+	}
+
+	protected void evaluateVariableDefinition(Group group, EvidenceVariable evidenceVariable, String subjectType, String subjectId) {
+		String definitionExpression = evidenceVariable.getCharacteristic().get(0).getDefinitionByCombination().getCharacteristic().get(0).getDefinitionExpression().getExpression();
+		if (definitionExpression != null && !definitionExpression.isEmpty()) {
+			Object result = this.evaluateDefinitionExpression(definitionExpression);
+			if (result instanceof Boolean) {
+				if (Boolean.TRUE.equals(result)) {
+					group.addMember().setEntity(new Reference(String.format("%s/%s", subjectType, subjectId)));
```

Comment:
For now we do not have re-indexing of the patient ? (it should probably be a configuration ?)